File: utils/integration_test.py
"""
End-to-end integration test for the Proof-of-Intent Protocol.

Requires a local Anvil node:
    anvil --port 8545

Run with:
    python -m utils.integration_test

The test deploys all four contracts plus a MockERC20 and MockSwapRouter from the
Foundry build artifacts, registers agents, runs the full delegation-and-execution
flow (including real tokenIn transferFrom + router call), then confirms that a
scope-exceeding delegation is correctly rejected.
"""

import json
import pathlib
import time

from eth_account import Account
from web3 import Web3
from web3.exceptions import ContractLogicError

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

_ROOT = pathlib.Path(__file__).parent.parent
_OUT  = _ROOT / "contracts" / "out"

TOKEN_OUT = "0x0000000000000000000000000000000000000001"
CHAIN_ID  = 31337  # Anvil default


def _load(contract_name: str) -> dict:
    path = _OUT / f"{contract_name}.sol" / f"{contract_name}.json"
    return json.loads(path.read_text())


def _deploy(w3, artifact, deployer, *constructor_args):
    abi      = artifact["abi"]
    bytecode = artifact["bytecode"]["object"]
    contract = w3.eth.contract(abi=abi, bytecode=bytecode)
    tx_hash  = contract.constructor(*constructor_args).transact({"from": deployer})
    receipt  = w3.eth.wait_for_transaction_receipt(tx_hash)
    assert receipt.status == 1, "Deploy failed"
    return w3.eth.contract(address=receipt.contractAddress, abi=abi)


def _b32(hex_str: str) -> bytes:
    s = hex_str[2:] if hex_str.startswith("0x") else hex_str
    return bytes.fromhex(s)


def _sign_intent(w3, intent: dict, private_key: str, registry_address: str) -> bytes:
    """Produce an EIP-712 signature for an Intent struct."""
    from eth_account.messages import encode_typed_data

    domain = {
        "name": "IntentRegistry",
        "version": "1",
        "chainId": CHAIN_ID,
        "verifyingContract": registry_address,
    }
    intent_type = {
        "Intent": [
            {"name": "owner",                  "type": "address"},
            {"name": "authorizedOrchestrator", "type": "address"},
            {"name": "tokenIn",                "type": "address"},
            {"name": "maxAmountIn",            "type": "uint256"},
            {"name": "minAmountOut",           "type": "uint256"},
            {"name": "allowedProtocols",       "type": "bytes32[]"},
            {"name": "deadline",               "type": "uint256"},
            {"name": "nonce",                  "type": "uint256"},
        ]
    }
    signed = Account.sign_typed_data(private_key, domain, intent_type, intent)
    return signed.signature


def _protocol_id(name: str) -> bytes:
    return Web3.keccak(text=name)


# ---------------------------------------------------------------------------
# Main integration test
# ---------------------------------------------------------------------------

def run():
    w3 = Web3(Web3.HTTPProvider("http://localhost:8545"))
    assert w3.is_connected(), "Anvil node not reachable at http://localhost:8545"

    accounts = w3.eth.accounts
    deployer           = accounts[0]
    user_key           = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"  # Anvil key #0
    orchestrator_key   = "0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d"  # Anvil key #1
    research_key       = "0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a"  # Anvil key #2
    execution_key      = "0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6"  # Anvil key #3

    user_addr          = Account.from_key(user_key).address
    orchestrator_addr  = Account.from_key(orchestrator_key).address
    research_addr      = Account.from_key(research_key).address
    execution_addr     = Account.from_key(execution_key).address

    print(f"User:         {user_addr}")
    print(f"Orchestrator: {orchestrator_addr}")
    print(f"Research:     {research_addr}")
    print(f"Execution:    {execution_addr}")

    # -------------------------------------------------------------------------
    # Deploy contracts
    # -------------------------------------------------------------------------
    print("\n[1] Deploying contracts...")

    agent_art      = _load("AgentRegistry")
    intent_art     = _load("IntentRegistry")
    deleg_art      = _load("DelegationRegistry")
    gate_art       = _load("ExecutionGate")
    mock_erc20_art = _load("MockERC20")
    mock_router_art = _load("MockSwapRouter")

    agent_reg   = _deploy(w3, agent_art, deployer)
    intent_reg  = _deploy(w3, intent_art, deployer)
    deleg_reg   = _deploy(w3, deleg_art, deployer, intent_reg.address, agent_reg.address)
    mock_token  = _deploy(w3, mock_erc20_art, deployer)
    mock_router = _deploy(w3, mock_router_art, deployer)
    exec_gate   = _deploy(w3, gate_art, deployer, intent_reg.address, deleg_reg.address, mock_router.address)

    TOKEN_IN = mock_token.address

    # Wire up execution gate
    tx = deleg_reg.functions.setExecutionGate(exec_gate.address).transact({"from": deployer})
    w3.eth.wait_for_transaction_receipt(tx)

    print(f"  AgentRegistry:      {agent_reg.address}")
    print(f"  IntentRegistry:     {intent_reg.address}")
    print(f"  DelegationRegistry: {deleg_reg.address}")
    print(f"  MockERC20 (tokenIn):{mock_token.address}")
    print(f"  MockSwapRouter:     {mock_router.address}")
    print(f"  ExecutionGate:      {exec_gate.address}")

    # -------------------------------------------------------------------------
    # Register agents
    # -------------------------------------------------------------------------
    print("\n[2] Registering agents...")

    for addr, name in [
        (orchestrator_addr, "Orchestrator"),
        (research_addr,     "ResearchAgent"),
        (execution_addr,    "ExecutionAgent"),
    ]:
        tx = agent_reg.functions.registerAgent(addr, name).transact({"from": deployer})
        w3.eth.wait_for_transaction_receipt(tx)
        print(f"  Registered {name}: {addr}")

    # -------------------------------------------------------------------------
    # Build and sign intent  (500 USDC-equivalent, Uniswap-V3, 60 min)
    # -------------------------------------------------------------------------
    print("\n[3] Building and signing intent...")

    protocol_id    = _protocol_id("Uniswap-V3")
    deadline       = int(time.time()) + 3600
    nonce          = intent_reg.functions.nonces(user_addr).call()
    max_amount_in  = 500 * 10**6          # 500 USDC (6 decimals)
    min_amount_out = 490 * 10**6

    intent_data = {
        "owner":                  user_addr,
        "authorizedOrchestrator": orchestrator_addr,
        "tokenIn":                Web3.to_checksum_address(TOKEN_IN),
        "maxAmountIn":            max_amount_in,
        "minAmountOut":           min_amount_out,
        "allowedProtocols":       [protocol_id],
        "deadline":               deadline,
        "nonce":                  nonce,
    }

    signature = _sign_intent(w3, intent_data, user_key, intent_reg.address)

    # -------------------------------------------------------------------------
    # Register intent
    # -------------------------------------------------------------------------
    print("\n[4] Registering intent on-chain...")

    intent_tuple = (
        user_addr,
        orchestrator_addr,
        Web3.to_checksum_address(TOKEN_IN),
        max_amount_in,
        min_amount_out,
        [protocol_id],
        deadline,
        nonce,
    )

    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(user_key).sign_transaction(
            intent_reg.functions.registerIntent(intent_tuple, signature).build_transaction({
                "from":    user_addr,
                "nonce":   w3.eth.get_transaction_count(user_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    assert receipt.status == 1, "registerIntent reverted"

    logs = intent_reg.events.IntentRegistered().process_receipt(receipt)
    intent_id = "0x" + logs[0]["args"]["intentId"].hex()
    print(f"  intentId: {intent_id}")

    # -------------------------------------------------------------------------
    # Delegate from root  (orchestrator → research agent, 400 USDC)
    # -------------------------------------------------------------------------
    print("\n[5] Delegating from root to research agent...")

    scope1 = (
        400 * 10**6,          # maxAmountIn: 400 USDC
        492 * 10**6,          # minAmountOut (tighter than root)
        [protocol_id],
        int(time.time()) + 1800,
    )

    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(orchestrator_key).sign_transaction(
            deleg_reg.functions.delegateFromRoot(
                _b32(intent_id), scope1, research_addr
            ).build_transaction({
                "from":    orchestrator_addr,
                "nonce":   w3.eth.get_transaction_count(orchestrator_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    assert receipt.status == 1, "delegateFromRoot reverted"

    logs = deleg_reg.events.DelegationCreated().process_receipt(receipt)
    delegation_id_1 = "0x" + logs[0]["args"]["delegationId"].hex()
    print(f"  delegation1 (root→research): {delegation_id_1}")

    # -------------------------------------------------------------------------
    # Delegate from delegation  (research → execution agent, 300 USDC)
    # -------------------------------------------------------------------------
    print("\n[6] Delegating from delegation to execution agent...")

    scope2 = (
        300 * 10**6,          # maxAmountIn: 300 USDC (narrower)
        494 * 10**6,          # minAmountOut (even tighter)
        [protocol_id],
        int(time.time()) + 900,
    )

    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(research_key).sign_transaction(
            deleg_reg.functions.delegateFromDelegation(
                _b32(delegation_id_1), scope2, execution_addr
            ).build_transaction({
                "from":    research_addr,
                "nonce":   w3.eth.get_transaction_count(research_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    assert receipt.status == 1, "delegateFromDelegation reverted"

    logs = deleg_reg.events.DelegationCreated().process_receipt(receipt)
    delegation_id_2 = "0x" + logs[0]["args"]["delegationId"].hex()
    print(f"  delegation2 (research→execution): {delegation_id_2}")

    # -------------------------------------------------------------------------
    # Verify chain (view call, must return True)
    # -------------------------------------------------------------------------
    print("\n[7] Verifying chain...")

    amount_in = 250 * 10**6   # 250 USDC (<= all scopes)
    tx_params = (
        amount_in,
        495 * 10**6,          # minAmountOut (>= all scopes)
        protocol_id,
        Web3.to_checksum_address(TOKEN_IN),
        Web3.to_checksum_address(TOKEN_OUT),
        execution_addr,       # recipient — also the tx sender
    )

    ok = exec_gate.functions.verifyChain(_b32(delegation_id_2), tx_params).call()
    assert ok, "verifyChain returned False"
    print("  verifyChain → True  ✓")

    # -------------------------------------------------------------------------
    # Mint tokenIn to execution_addr (recipient) and approve ExecutionGate
    # -------------------------------------------------------------------------
    print("\n[7b] Setting up token approval for executeSwap...")

    # Mint enough tokenIn to the recipient (execution_addr acts as both caller
    # and recipient in this integration test).
    tx = mock_token.functions.mint(execution_addr, amount_in * 2).transact({"from": deployer})
    w3.eth.wait_for_transaction_receipt(tx)

    # execution_addr approves ExecutionGate to pull its tokenIn.
    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(execution_key).sign_transaction(
            mock_token.functions.approve(exec_gate.address, amount_in * 2).build_transaction({
                "from":    execution_addr,
                "nonce":   w3.eth.get_transaction_count(execution_addr),
                "chainId": CHAIN_ID,
                "gas":     100_000,
            })
        ).raw_transaction
    )
    w3.eth.wait_for_transaction_receipt(tx_hash)
    print(f"  Minted {amount_in} tokenIn to {execution_addr}")
    print(f"  Approved ExecutionGate ({exec_gate.address}) to spend tokenIn  ✓")

    # -------------------------------------------------------------------------
    # Execute swap
    # -------------------------------------------------------------------------
    print("\n[8] Calling executeSwap...")

    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(execution_key).sign_transaction(
            exec_gate.functions.executeSwap(
                _b32(delegation_id_2), tx_params
            ).build_transaction({
                "from":    execution_addr,
                "nonce":   w3.eth.get_transaction_count(execution_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    assert receipt.status == 1, "executeSwap reverted"

    delegation_state = deleg_reg.functions.getDelegation(_b32(delegation_id_2)).call()
    assert delegation_state[4], "executed flag not set after executeSwap"
    print(f"  executeSwap tx: 0x{receipt['transactionHash'].hex()}")
    print("  delegation.executed → True  ✓")

    # -------------------------------------------------------------------------
    # Malicious delegation: 800 USDC when only 400 was authorized (scope1)
    # This must be rejected with "Amount exceeds scope"
    # -------------------------------------------------------------------------
    print("\n[9] Attempting malicious delegation (800 USDC > 400 authorized)...")

    # Need a fresh root delegation on a new intent (replay protection consumed scope1).
    # Register a second intent (nonce = 1) and delegate from it.
    nonce2 = intent_reg.functions.nonces(user_addr).call()
    intent_data2 = {**intent_data, "nonce": nonce2}
    sig2 = _sign_intent(w3, intent_data2, user_key, intent_reg.address)
    intent_tuple2 = (
        user_addr, orchestrator_addr, Web3.to_checksum_address(TOKEN_IN),
        max_amount_in, min_amount_out, [protocol_id], deadline, nonce2,
    )
    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(user_key).sign_transaction(
            intent_reg.functions.registerIntent(intent_tuple2, sig2).build_transaction({
                "from":    user_addr,
                "nonce":   w3.eth.get_transaction_count(user_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt2 = w3.eth.wait_for_transaction_receipt(tx_hash)
    logs2 = intent_reg.events.IntentRegistered().process_receipt(receipt2)
    intent_id2 = "0x" + logs2[0]["args"]["intentId"].hex()

    # Delegate from root: authorize research with 400 USDC
    scope_limited = (400 * 10**6, 492 * 10**6, [protocol_id], int(time.time()) + 1800)
    tx_hash = w3.eth.send_raw_transaction(
        Account.from_key(orchestrator_key).sign_transaction(
            deleg_reg.functions.delegateFromRoot(
                _b32(intent_id2), scope_limited, research_addr
            ).build_transaction({
                "from":    orchestrator_addr,
                "nonce":   w3.eth.get_transaction_count(orchestrator_addr),
                "chainId": CHAIN_ID,
                "gas":     500_000,
            })
        ).raw_transaction
    )
    receipt3 = w3.eth.wait_for_transaction_receipt(tx_hash)
    logs3 = deleg_reg.events.DelegationCreated().process_receipt(receipt3)
    delegation_limited = "0x" + logs3[0]["args"]["delegationId"].hex()

    # Now research tries to sub-delegate 800 USDC — must revert.
    # Use eth_call to get the revert reason before broadcasting.
    scope_malicious = (800 * 10**6, 492 * 10**6, [protocol_id], int(time.time()) + 900)
    revert_reason = ""
    reverted = False
    try:
        deleg_reg.functions.delegateFromDelegation(
            _b32(delegation_limited), scope_malicious, execution_addr
        ).call({"from": research_addr})
    except ContractLogicError as exc:
        reverted = True
        revert_reason = str(exc)
    except Exception as exc:
        reverted = True
        revert_reason = str(exc)

    assert reverted, "Expected revert for malicious delegation but got none"
    assert "Amount exceeds scope" in revert_reason, (
        f"Expected 'Amount exceeds scope' in revert reason, got: {revert_reason!r}"
    )
    print(f"  Malicious delegation correctly reverted with: {revert_reason[:80]}")
    print("  'Amount exceeds scope'  ✓")

    print("\n========================================")
    print("All integration tests passed.")
    print("========================================")


if __name__ == "__main__":
    run()

File: intent-custody-ts/test/integrationTest.ts
/**
 * End-to-end integration test for the Proof-of-Intent Protocol.
 *
 * Requires a local Anvil node:
 *     anvil --port 8545
 *
 * Run with:
 *     npm test
 *
 * The test deploys all four contracts plus a MockERC20 and MockSwapRouter from
 * the Foundry build artifacts, registers agents, runs the full delegation-and-
 * execution flow (including real tokenIn transferFrom + router call), then
 * confirms that a scope-exceeding delegation is correctly rejected.
 */

import { ethers, NonceManager } from 'ethers';
import { readFileSync } from 'fs';
import { join } from 'path';

const CHAIN_ID = 31337; // Anvil default
const TOKEN_OUT = '0x0000000000000000000000000000000000000001';

// Anvil default private keys
const DEPLOYER_KEY      = '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'; // key #0
const ORCHESTRATOR_KEY  = '0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d'; // key #1
const RESEARCH_KEY      = '0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a'; // key #2
const EXECUTION_KEY     = '0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6'; // key #3

const OUT = join(__dirname, '../../contracts/out');

function loadArtifact(contractName: string): { abi: unknown[]; bytecode: string } {
  const raw = JSON.parse(
    readFileSync(join(OUT, `${contractName}.sol`, `${contractName}.json`), 'utf8'),
  ) as { abi: unknown[]; bytecode: { object: string } };
  return { abi: raw.abi, bytecode: raw.bytecode.object };
}

async function deploy(
  wallet: ethers.Signer,
  artifact: { abi: unknown[]; bytecode: string },
  ...args: unknown[]
): Promise<ethers.Contract> {
  const factory = new ethers.ContractFactory(artifact.abi, artifact.bytecode, wallet);
  const contract = await factory.deploy(...args);
  await contract.waitForDeployment();
  return contract as ethers.Contract;
}

function protocolId(name: string): string {
  return ethers.id(name);
}

async function signIntent(
  intent: Record<string, unknown>,
  wallet: ethers.Signer,
  registryAddress: string,
): Promise<string> {
  const domain = {
    name: 'IntentRegistry',
    version: '1',
    chainId: CHAIN_ID,
    verifyingContract: registryAddress,
  };
  const types = {
    Intent: [
      { name: 'owner',                  type: 'address' },
      { name: 'authorizedOrchestrator', type: 'address' },
      { name: 'tokenIn',                type: 'address' },
      { name: 'maxAmountIn',            type: 'uint256' },
      { name: 'minAmountOut',           type: 'uint256' },
      { name: 'allowedProtocols',       type: 'bytes32[]' },
      { name: 'deadline',               type: 'uint256' },
      { name: 'nonce',                  type: 'uint256' },
    ],
  };
  return wallet.signTypedData(domain, types, intent);
}

function parseEventArg(
  receipt: ethers.TransactionReceipt,
  iface: ethers.Interface,
  eventName: string,
  argName: string,
): string {
  for (const log of receipt.logs) {
    try {
      const parsed = iface.parseLog({ topics: [...log.topics], data: log.data });
      if (parsed?.name === eventName) return parsed.args[argName] as string;
    } catch { /* log belongs to a different contract */ }
  }
  throw new Error(`${eventName} event not found in receipt`);
}

async function run(): Promise<void> {
  const provider = new ethers.JsonRpcProvider('http://localhost:8545');
  try {
    await provider.getNetwork();
  } catch {
    throw new Error('Anvil node not reachable at http://localhost:8545');
  }

  // Wrap each wallet in NonceManager so rapid sequential txs get correct nonces
  // without racing against eth_getTransactionCount returning stale values.
  const deployerWallet      = new NonceManager(new ethers.Wallet(DEPLOYER_KEY, provider));
  const orchestratorWallet  = new NonceManager(new ethers.Wallet(ORCHESTRATOR_KEY, provider));
  const researchWallet      = new NonceManager(new ethers.Wallet(RESEARCH_KEY, provider));
  const executionWallet     = new NonceManager(new ethers.Wallet(EXECUTION_KEY, provider));

  // In the Python test, deployer == accounts[0] == user (Anvil key #0)
  const userAddr         = await deployerWallet.getAddress();
  const orchestratorAddr = await orchestratorWallet.getAddress();
  const researchAddr     = await researchWallet.getAddress();
  const executionAddr    = await executionWallet.getAddress();

  console.log(`User:         ${userAddr}`);
  console.log(`Orchestrator: ${orchestratorAddr}`);
  console.log(`Research:     ${researchAddr}`);
  console.log(`Execution:    ${executionAddr}`);

  // ---------------------------------------------------------------------------
  // [1] Deploy contracts
  // ---------------------------------------------------------------------------
  console.log('\n[1] Deploying contracts...');

  const agentArt      = loadArtifact('AgentRegistry');
  const intentArt     = loadArtifact('IntentRegistry');
  const delegArt      = loadArtifact('DelegationRegistry');
  const gateArt       = loadArtifact('ExecutionGate');
  const mockERC20Art  = loadArtifact('MockERC20');
  const mockRouterArt = loadArtifact('MockSwapRouter');

  const agentReg   = await deploy(deployerWallet, agentArt);
  const intentReg  = await deploy(deployerWallet, intentArt);
  const mockToken  = await deploy(deployerWallet, mockERC20Art);
  const mockRouter = await deploy(deployerWallet, mockRouterArt);

  const agentRegAddr   = await agentReg.getAddress();
  const intentRegAddr  = await intentReg.getAddress();
  const mockTokenAddr  = await mockToken.getAddress();
  const mockRouterAddr = await mockRouter.getAddress();

  const delegReg = await deploy(deployerWallet, delegArt, intentRegAddr, agentRegAddr);
  const delegRegAddr = await delegReg.getAddress();

  const execGate = await deploy(deployerWallet, gateArt, intentRegAddr, delegRegAddr, mockRouterAddr);
  const execGateAddr = await execGate.getAddress();

  // Wire up execution gate
  const wireTx = await delegReg.setExecutionGate(execGateAddr) as ethers.ContractTransactionResponse;
  await wireTx.wait();

  console.log(`  AgentRegistry:      ${agentRegAddr}`);
  console.log(`  IntentRegistry:     ${intentRegAddr}`);
  console.log(`  DelegationRegistry: ${delegRegAddr}`);
  console.log(`  MockERC20 (tokenIn):${mockTokenAddr}`);
  console.log(`  MockSwapRouter:     ${mockRouterAddr}`);
  console.log(`  ExecutionGate:      ${execGateAddr}`);

  // ---------------------------------------------------------------------------
  // [2] Register agents
  // ---------------------------------------------------------------------------
  console.log('\n[2] Registering agents...');

  for (const [addr, name] of [
    [orchestratorAddr, 'Orchestrator'],
    [researchAddr,     'ResearchAgent'],
    [executionAddr,    'ExecutionAgent'],
  ] as [string, string][]) {
    const tx = await agentReg.registerAgent(addr, name) as ethers.ContractTransactionResponse;
    await tx.wait();
    console.log(`  Registered ${name}: ${addr}`);
  }

  // ---------------------------------------------------------------------------
  // [3] Build and sign intent (500 USDC-equivalent, Uniswap-V3, 60 min)
  // ---------------------------------------------------------------------------
  console.log('\n[3] Building and signing intent...');

  const uniswapV3Id  = protocolId('Uniswap-V3');
  const deadline     = BigInt(Math.floor(Date.now() / 1000) + 3600);
  const nonce        = (await intentReg.nonces(userAddr)) as bigint;
  const maxAmountIn  = BigInt(500 * 10 ** 6);   // 500 USDC (6 decimals)
  const minAmountOut = BigInt(490 * 10 ** 6);

  const intentData = {
    owner:                  userAddr,
    authorizedOrchestrator: orchestratorAddr,
    tokenIn:                ethers.getAddress(mockTokenAddr),
    maxAmountIn,
    minAmountOut,
    allowedProtocols:       [uniswapV3Id],
    deadline,
    nonce,
  };

  const signature = await signIntent(intentData, deployerWallet, intentRegAddr);

  // ---------------------------------------------------------------------------
  // [4] Register intent
  // ---------------------------------------------------------------------------
  console.log('\n[4] Registering intent on-chain...');

  const intentTuple = [
    userAddr,
    orchestratorAddr,
    ethers.getAddress(mockTokenAddr),
    maxAmountIn,
    minAmountOut,
    [uniswapV3Id],
    deadline,
    nonce,
  ];

  const regTx = await (intentReg.connect(deployerWallet) as ethers.Contract).registerIntent(
    intentTuple, signature,
  ) as ethers.ContractTransactionResponse;
  const regReceipt = await regTx.wait();
  if (!regReceipt || regReceipt.status !== 1) throw new Error('registerIntent reverted');

  const intentId = parseEventArg(regReceipt, intentReg.interface, 'IntentRegistered', 'intentId');
  console.log(`  intentId: ${intentId}`);

  // ---------------------------------------------------------------------------
  // [5] Delegate from root (orchestrator → research agent, 400 USDC)
  // ---------------------------------------------------------------------------
  console.log('\n[5] Delegating from root to research agent...');

  const scope1 = [
    BigInt(400 * 10 ** 6),
    BigInt(492 * 10 ** 6),
    [uniswapV3Id],
    BigInt(Math.floor(Date.now() / 1000) + 1800),
  ];

  const deleg1Tx = await (delegReg.connect(orchestratorWallet) as ethers.Contract).delegateFromRoot(
    intentId, scope1, researchAddr,
  ) as ethers.ContractTransactionResponse;
  const deleg1Receipt = await deleg1Tx.wait();
  if (!deleg1Receipt || deleg1Receipt.status !== 1) throw new Error('delegateFromRoot reverted');

  const delegationId1 = parseEventArg(deleg1Receipt, delegReg.interface, 'DelegationCreated', 'delegationId');
  console.log(`  delegation1 (root→research): ${delegationId1}`);

  // ---------------------------------------------------------------------------
  // [6] Delegate from delegation (research → execution agent, 300 USDC)
  // ---------------------------------------------------------------------------
  console.log('\n[6] Delegating from delegation to execution agent...');

  const scope2 = [
    BigInt(300 * 10 ** 6),
    BigInt(494 * 10 ** 6),
    [uniswapV3Id],
    BigInt(Math.floor(Date.now() / 1000) + 900),
  ];

  const deleg2Tx = await (delegReg.connect(researchWallet) as ethers.Contract).delegateFromDelegation(
    delegationId1, scope2, executionAddr,
  ) as ethers.ContractTransactionResponse;
  const deleg2Receipt = await deleg2Tx.wait();
  if (!deleg2Receipt || deleg2Receipt.status !== 1) throw new Error('delegateFromDelegation reverted');

  const delegationId2 = parseEventArg(deleg2Receipt, delegReg.interface, 'DelegationCreated', 'delegationId');
  console.log(`  delegation2 (research→execution): ${delegationId2}`);

  // ---------------------------------------------------------------------------
  // [7] Verify chain (view call, must return true)
  // ---------------------------------------------------------------------------
  console.log('\n[7] Verifying chain...');

  const amountIn = BigInt(250 * 10 ** 6); // 250 USDC (<= all scopes)
  const txParams = [
    amountIn,
    BigInt(495 * 10 ** 6), // minAmountOut (>= all scopes)
    uniswapV3Id,
    ethers.getAddress(mockTokenAddr),
    ethers.getAddress(TOKEN_OUT),
    executionAddr, // recipient — also the tx sender
  ];

  const ok = await execGate.verifyChain(delegationId2, txParams) as boolean;
  if (!ok) throw new Error('verifyChain returned false');
  console.log('  verifyChain → True  ✓');

  // ---------------------------------------------------------------------------
  // [7b] Mint tokenIn to executionAddr and approve ExecutionGate
  // ---------------------------------------------------------------------------
  console.log('\n[7b] Setting up token approval for executeSwap...');

  const mintTx = await (mockToken.connect(deployerWallet) as ethers.Contract).mint(
    executionAddr, amountIn * 2n,
  ) as ethers.ContractTransactionResponse;
  await mintTx.wait();

  const approveTx = await (mockToken.connect(executionWallet) as ethers.Contract).approve(
    execGateAddr, amountIn * 2n,
  ) as ethers.ContractTransactionResponse;
  await approveTx.wait();

  console.log(`  Minted ${amountIn} tokenIn to ${executionAddr}`);
  console.log(`  Approved ExecutionGate (${execGateAddr}) to spend tokenIn  ✓`);

  // ---------------------------------------------------------------------------
  // [8] Execute swap
  // ---------------------------------------------------------------------------
  console.log('\n[8] Calling executeSwap...');

  const swapTx = await (execGate.connect(executionWallet) as ethers.Contract).executeSwap(
    delegationId2, txParams,
  ) as ethers.ContractTransactionResponse;
  const swapReceipt = await swapTx.wait();
  if (!swapReceipt || swapReceipt.status !== 1) throw new Error('executeSwap reverted');

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const delegState = (await delegReg.getDelegation(delegationId2)) as any;
  if (!delegState.executed) throw new Error('executed flag not set after executeSwap');
  console.log(`  executeSwap tx: ${swapReceipt.hash}`);
  console.log('  delegation.executed → True  ✓');

  // ---------------------------------------------------------------------------
  // [9] Malicious delegation: 800 USDC when only 400 was authorized (scope1).
  //     This must be rejected with "Amount exceeds scope".
  // Need a fresh root delegation on a new intent (replay protection consumed scope1).
  // ---------------------------------------------------------------------------
  console.log('\n[9] Attempting malicious delegation (800 USDC > 400 authorized)...');

  const nonce2 = (await intentReg.nonces(userAddr)) as bigint;
  const intentData2 = { ...intentData, nonce: nonce2 };
  const sig2 = await signIntent(intentData2, deployerWallet, intentRegAddr);
  const intentTuple2 = [
    userAddr, orchestratorAddr, ethers.getAddress(mockTokenAddr),
    maxAmountIn, minAmountOut, [uniswapV3Id], deadline, nonce2,
  ];

  const reg2Tx = await (intentReg.connect(deployerWallet) as ethers.Contract).registerIntent(
    intentTuple2, sig2,
  ) as ethers.ContractTransactionResponse;
  const reg2Receipt = await reg2Tx.wait();
  const intentId2 = parseEventArg(reg2Receipt!, intentReg.interface, 'IntentRegistered', 'intentId');

  const scopeLimited = [
    BigInt(400 * 10 ** 6),
    BigInt(492 * 10 ** 6),
    [uniswapV3Id],
    BigInt(Math.floor(Date.now() / 1000) + 1800),
  ];

  const deleg3Tx = await (delegReg.connect(orchestratorWallet) as ethers.Contract).delegateFromRoot(
    intentId2, scopeLimited, researchAddr,
  ) as ethers.ContractTransactionResponse;
  const deleg3Receipt = await deleg3Tx.wait();
  const delegationLimited = parseEventArg(
    deleg3Receipt!, delegReg.interface, 'DelegationCreated', 'delegationId',
  );

  // Research tries to sub-delegate 800 USDC — must revert.
  // Use staticCall to get the revert reason without broadcasting.
  const scopeMalicious = [
    BigInt(800 * 10 ** 6),
    BigInt(492 * 10 ** 6),
    [uniswapV3Id],
    BigInt(Math.floor(Date.now() / 1000) + 900),
  ];

  let reverted = false;
  let revertReason = '';
  try {
    await (delegReg.connect(researchWallet) as ethers.Contract).delegateFromDelegation.staticCall(
      delegationLimited, scopeMalicious, executionAddr,
    );
  } catch (err: unknown) {
    reverted = true;
    revertReason = err instanceof Error ? err.message : String(err);
  }

  if (!reverted) throw new Error('Expected revert for malicious delegation but got none');
  if (!revertReason.includes('Amount exceeds scope')) {
    throw new Error(`Expected 'Amount exceeds scope' in revert reason, got: ${revertReason}`);
  }
  console.log(`  Malicious delegation correctly reverted with: ${revertReason.slice(0, 80)}`);
  console.log("  'Amount exceeds scope'  ✓");

  console.log('\n========================================');
  console.log('All integration tests passed.');
  console.log('========================================');
}

run().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
